### PR TITLE
Add bulk importing guide

### PR DESCRIPTION
**Pull request by jamiltz**
Connects to #390

### CODE REVIEW

File: md-docs/ready/guides/sync-gateway/bulk-importing.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+To prevent this behaviour you could map the `id` property of each movie to `_id` before sending the `/{db}/_bulk_docs` request. When you don't have control over the external API those data transformations are common practice.
+
+## Keeping two data sources in sync
+
+There are scenarios where the data from the stub API may change over time (for example the release date of a movie). In this case, the `_bulk_docs` would return a `409` status in the response body indicating that a document with this `_id` already exists in the database. At that point you need to do a read (to retrieve the revision number) before you can attempt to write the document again.
+
+Furthermore, there could be a scenario where the document inserted from the stub API can also be updated by users (if the user reviews were embedded in the movie document for example). In this case, it becomes  challenging to keep them in sync, it would involve doing 3-way merges to keep the user and stub API updates in the same document. For this reason, we recommend that all data inserted from an external source should be immutable once inserted in Sync Gateway (documents that cannot be updated).
```
Comment (adamcfraser) [inline md-docs/ready/guides/sync-gateway/bulk-importing.md]: Do we actually recommend that data inserted from an external source should be immutable?  That seems like an extreme recommendation, and depends a lot on the use case.  I guess it might be true for data that's intended to be continuously updated from an external source (i.e. you've got an ongoing migration process that's updating the doc).

It's also not clear from the discussion how that would be done (making the document immutable).

**Comment by jamiltz** [comment]
@jens the section of the guide on bulk importing is https://github.com/couchbaselabs/couchbase-mobile-portal/pull/392/files#diff-0ef51c9d82837a43b9f442d01c8e3de7R84, we could add more detail in the paragraph about "keeping two data sources in sync" but i don’t know what direction to take. Or keep it as an overview and come back to it later.

File: md-docs/ready/guides/sync-gateway/bulk-importing.md (hunk 1 above)
Comment (jamiltz) [inline md-docs/ready/guides/sync-gateway/bulk-importing.md]: That was the conclusion I came to after struggling to write the code to manage the scenario where both sides can update the data.

I was working from the assumption that on the external doc, there is no `update_at` or revision identifier and that's why the 3 way merge would be needed.

But it looks like we should assume the external doc has this form of identification to know if there was an update since the last time it was imported to SG. If there was an update (i.e the external revision number in SG is different to the latest external rev in the external API) then I'm still wondering how to not overwrite updates made by users in the Couchbase Mobile world.

**Comment by jamiltz** [comment]
@snej Per the discussion during standup last week, can you provide your input on how to manage updates from two data sources, updates in Couchbase Lite & Sync Gateway and the ones in the "external system".

Should the guide assume that the external API has an update identifier similar to the revision number in CBMobile?

**Comment by snej** [comment]
This is really similar to how bucket shadowing works, and the source code for shadowing is actually [pretty small](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go) and hopefully not too hard to follow.

To [pull from an external store](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L95) you first need to establish some sort of feed that notifies you when a record in the external store gets added/changed/deleted. You then read the record and convert it to JSON if necessary.

If there's any chance you might ever get notified of the same change twice (if the feed might contain dups, or if you ever have to reread an old portion of the feed), then get an identifier corresponding to the exact version of the document in the feed, like a CAS value or a timestamp or clock, and associate it with the document ID. (The bucket shadower calls this `UpstreamCAS`.) Then if you ever get a record with the same value again, [ignore it](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L113-L115).

If the new record contents are identical to the doc's current revision, we assume it's an echo of a change pushed from Couchbase Mobile; in that case we [don't save it as a new revision](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L121), but we still get its CAS/clock value and save that as the `UpstreamCAS`. (You can ignore this check if you don't push anything to the external store.)

We also keep an `UpstreamRev` property associated with the document, which identifies the Couchbase Mobile revID associated with the latest external record. When we save the changed record to CBM, we make it a child of that (by [setting the `_rev` property to `UpstreamRev`](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L126-L128) before saving.) _This may create a conflict,_ but that's fine — this represents the doc being changed at the same time both in the external store and in Couchbase, and the conflict can be resolved as usual in CBL or SG.

[Pushing from Couchbase Mobile](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L154) to an external store is really easy: you just watch the changes feed, and copy the document's current revision to the store. But first [compare the new revision's ID to the associated `UpstreamRev`](https://github.com/couchbase/sync_gateway/blob/master/db/shadower.go#L158-L160) — if they're equal, this is an echo of a record pulled from the external source, and you should ignore it.

**Comment by jamiltz** [comment]
Thanks!!

Updated outline:
- Introduction
- Stub API
- Keeping two data sources in sync
  - _Importing documents_: the [stub API](https://cl.ly/272h2r1M243x/movies.json) doesn't have a feed to get notified of changes when they happen. So the guide will assume that the imports from the external source to CBM happen periodically. I will add a note about connecting to the external feed if available on the external source.
  - _Exporting documents_: I'll write this section that shows how to consume the changes feed and write to the external source. The stub API doesn't allow write operation currently. To add this capability and keeps things simple, we could add a small web server with one POST endpoint that, given a doc id, will update it in movies.json (the external source).
- Importing attachments


So in the "Keeping two data sources in sync" section, we assume imports from external source to CBM are periodic. And updates from CBM to external source are in real-time via the changes feed. That ensures that the external doc always gets updated with CBM changes so it won't overwrite CBM updates once the doc is updated on the external system and written back to CBM.
